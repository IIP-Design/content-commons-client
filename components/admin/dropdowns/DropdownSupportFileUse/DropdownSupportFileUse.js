import React from 'react';
import PropTypes from 'prop-types';
import { Dropdown } from 'semantic-ui-react';
import gql from 'graphql-tag';
import { graphql } from 'react-apollo';

const DropdownSupportFileUse = props => {
  const { data, ...rest } = props;
  if ( data.loading ) return 'Loading file use...';
  if ( data.error ) return 'Loading error...';
  return <Dropdown { ...rest } options={ data.uses } selection />;
};

DropdownSupportFileUse.propTypes = {
  id: PropTypes.string,
  name: PropTypes.string,
  onChange: PropTypes.func,
  value: PropTypes.string,
  data: PropTypes.object,
  fluid: PropTypes.bool,
  required: PropTypes.bool
};

const SUPPORT_FILE_USES_QUERY = gql`
  query SupportFileUse($orderBy: SupportFileUseOrderByInput) {
    uses: supportFileUses(orderBy: $orderBy) {
      value: id
      text: name
    }
  }
`;

export default graphql( SUPPORT_FILE_USES_QUERY, {
  options: {
    variables: { orderBy: 'name_ASC' }
  }
} )( DropdownSupportFileUse );

export { SUPPORT_FILE_USES_QUERY };
