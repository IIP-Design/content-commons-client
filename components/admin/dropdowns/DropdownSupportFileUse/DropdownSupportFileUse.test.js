import { mount } from 'enzyme';
import toJSON from 'enzyme-to-json';
import wait from 'waait';
import { MockedProvider } from 'react-apollo/test-utils';
import DropdownSupportFileUse, { SUPPORT_FILE_USES_QUERY } from './DropdownSupportFileUse';

const props = {
  id: 'm123',
  name: 'fileUse',
  onChange: jest.fn(),
  value: 'th89',
  fluid: true,
  required: true
};

const mocks = [
  {
    request: {
      query: SUPPORT_FILE_USES_QUERY,
      variables: { orderBy: 'name_ASC' }
    },
    result: {
      data: {
        uses: [
          { value: 'th89', text: 'Thumbnail/Cover Image' },
          { value: 'em92', text: 'Email Graphic' }
        ]
      }
    }
  }
];

const Component = (
  <MockedProvider mocks={ mocks } addTypename={ false }>
    <DropdownSupportFileUse { ...props } />
  </MockedProvider>
);

describe( '<DropdownSupportFileUse />', () => {
  it( 'renders initial loading state without crashing', () => {
    const wrapper = mount( Component );
    const dropdown = wrapper.find( 'DropdownSupportFileUse' );

    expect( dropdown.exists() ).toEqual( true );
    expect( dropdown.contains( 'Loading file use...' ) ).toEqual( true );
  } );

  it( 'renders error message & icon if error is thrown', async () => {
    const errorMocks = [
      {
        request: {
          query: SUPPORT_FILE_USES_QUERY,
          variables: { orderBy: 'name_ASC' }
        },
        result: {
          errors: [{ message: 'There was an error.' }]
        }
      }
    ];
    const wrapper = mount(
      <MockedProvider mocks={ errorMocks } addTypename={ false }>
        <DropdownSupportFileUse { ...props } />
      </MockedProvider>
    );
    // wait for the data and !loading
    await wait( 0 );
    wrapper.update();

    const dropdown = wrapper.find( 'DropdownSupportFileUse' );

    expect( dropdown.contains( 'Loading error...' ) ).toEqual( true );
  } );

  it( 'renders the dropdown with correct uses', async () => {
    const wrapper = mount( Component );
    await wait( 0 );
    wrapper.update();

    const dropdownUse = wrapper.find( 'DropdownSupportFileUse' );
    const dropdownSemanticUI = dropdownUse.find( 'Dropdown' );
    const renderedUseItems = dropdownSemanticUI.find( 'DropdownItem' );
    const { uses } = dropdownUse.prop( 'data' );

    expect( dropdownSemanticUI.exists() ).toEqual( true );
    expect( renderedUseItems.length ).toEqual( uses.length );
    renderedUseItems.forEach( ( item, i ) => {
      expect( item.prop( 'value' ) ).toEqual( uses[i].value );
      expect( item.prop( 'text' ) ).toEqual( uses[i].text );
    } );
    expect( toJSON( dropdownUse ) ).toMatchSnapshot();
  } );

  it( 'making a selection calls `onChange`', async () => {
    const wrapper = mount( Component );
    await wait( 0 );
    wrapper.update();

    const dropdown = wrapper.find( 'DropdownSupportFileUse' );

    dropdown.simulate( 'change' );
    expect( props.onChange ).toHaveBeenCalled();
  } );
} );
