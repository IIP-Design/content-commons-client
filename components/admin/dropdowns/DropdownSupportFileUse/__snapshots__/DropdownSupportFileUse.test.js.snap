// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DropdownSupportFileUse /> renders the dropdown with correct uses 1`] = `
<DropdownSupportFileUse
  data={
    Object {
      "error": undefined,
      "fetchMore": [Function],
      "loading": false,
      "networkStatus": 7,
      "refetch": [Function],
      "startPolling": [Function],
      "stopPolling": [Function],
      "subscribeToMore": [Function],
      "updateQuery": [Function],
      "uses": Array [
        Object {
          "text": "Thumbnail/Cover Image",
          "value": "th89",
        },
        Object {
          "text": "Email Graphic",
          "value": "em92",
        },
      ],
      "variables": Object {
        "orderBy": "name_ASC",
      },
    }
  }
  fluid={true}
  id="m123"
  name="fileUse"
  onChange={[MockFunction]}
  required={true}
  value="th89"
>
  <Dropdown
    additionLabel="Add "
    additionPosition="top"
    closeOnBlur={true}
    deburr={false}
    fluid={true}
    icon="dropdown"
    id="m123"
    minCharacters={1}
    name="fileUse"
    noResultsMessage="No results found."
    onChange={[MockFunction]}
    openOnFocus={true}
    options={
      Array [
        Object {
          "text": "Thumbnail/Cover Image",
          "value": "th89",
        },
        Object {
          "text": "Email Graphic",
          "value": "em92",
        },
      ]
    }
    renderLabel={[Function]}
    required={true}
    searchInput="text"
    selectOnBlur={true}
    selectOnNavigation={true}
    selection={true}
    value="th89"
    wrapSelection={true}
  >
    <div
      aria-expanded={false}
      className="ui fluid selection dropdown"
      id="m123"
      name="fileUse"
      onBlur={[Function]}
      onChange={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onMouseDown={[Function]}
      required={true}
      role="listbox"
      tabIndex={0}
    >
      <div
        aria-live="polite"
        className="text"
        role="alert"
      >
        Thumbnail/Cover Image
      </div>
      <Icon
        as="i"
        className=""
        name="dropdown"
        onClick={[Function]}
      >
        <i
          aria-hidden="true"
          className="dropdown icon"
          onClick={[Function]}
        />
      </Icon>
      <DropdownMenu>
        <div
          className="menu transition"
        >
          <DropdownItem
            active={true}
            key="th89"
            onClick={[Function]}
            selected={true}
            style={
              Object {
                "pointerEvents": "all",
              }
            }
            text="Thumbnail/Cover Image"
            value="th89"
          >
            <div
              aria-checked={true}
              aria-selected={true}
              className="active selected item"
              onClick={[Function]}
              role="option"
              style={
                Object {
                  "pointerEvents": "all",
                }
              }
            >
              <span
                className="text"
              >
                Thumbnail/Cover Image
              </span>
            </div>
          </DropdownItem>
          <DropdownItem
            active={false}
            key="em92"
            onClick={[Function]}
            selected={false}
            style={
              Object {
                "pointerEvents": "all",
              }
            }
            text="Email Graphic"
            value="em92"
          >
            <div
              aria-checked={false}
              aria-selected={false}
              className="item"
              onClick={[Function]}
              role="option"
              style={
                Object {
                  "pointerEvents": "all",
                }
              }
            >
              <span
                className="text"
              >
                Email Graphic
              </span>
            </div>
          </DropdownItem>
        </div>
      </DropdownMenu>
    </div>
  </Dropdown>
</DropdownSupportFileUse>
`;
